refactor(protocol-designer): various deck setup flex stacker feature

jerader opened the pull request:
closes EXEC-2121
# Overview

various deck setup touchups for flex stacker:
1. you can't drag/drop to/from the stacker (shuttle and hopper) for now anyway
2. selecting to add labware to the shuttle and hopper only allows the tiprack + lid or labware, the shuttle can add lids to the labware but the hopper cannot
3. you can edit the quantity on the hopper now
4. if you add a stack of tipracks + lids on the hopper, all the entities are now created
5. edits to the edit quanity modal to account for changing the number of labware + tiprack/lid combos
6. some fixes to the labware toolbox to edit the liquids in a stack -> NOTE this isn't fully done yet but i got tired so i'll do that in a follow up

## Test Plan and Hands on Testing

Stuffs, test the above items that i've edited

## Changelog

a bunch of refactoring for various deck setup components to accommodate the hopper and shuttle

## Risk assessment

low

jerader committed:
refactor(protocol-deisgner): prevent pipetting into shuttle and allow moving to shuttle

closes EXEC-2122

jerader committed:
filter out mix and transfer buttons

jerader committed:
fix lint

jerader committed:
rename const

jerader committed:
deck setup touchupsss

jerader committed:
stuff

jerader committed:
edit toolbox

jerader committed:
refactor(protocol-designer): various deck setup flex stacker features

closes EXEC-21221

jerader committed:
remove console.logs

jerader committed:
Merge branch 'edge' into pd_deck-setup-touchups

jerader committed:
fix merge conflict

jerader committed:
clean up lint

jerader committed:
fix selectAllLabware action in assignLiquidsModal

jerader committed:
Merge branch 'edge' into pd_deck-setup-touchups

jerader committed:
fix lint

ncdiehl11 commented on protocol-designer/src/components/organisms/AssignLiquidsModal/LabwareToolbox.tsx:
```diff
@@ -60,27 +60,32 @@ export function LabwareStackToolbox({
+    if (filteredLabwareStack.length < hopperStackLimit) {
+      dispatch(
+        createContainer({
+          labwareDefURIStack: [labware[labwareId ?? '']?.labwareDefURI ?? ''],
```
not intro'd here but maybe we should move this to the if statement on line 73 so we don't end up empty strings in the stack

ncdiehl11 commented on protocol-designer/src/components/organisms/EditLabwareQuantityModal/index.tsx:
```diff
@@ -25,24 +25,39 @@ import { getLabwareEntities } from '/protocol-designer/step-forms/selectors'
+  const zHeight = def.dimensions.zDimension
+  const stackingLimit = isOnHopper
+    ? getHopperStackLimit(zHeight)
+    : (def.stackLimit ?? 0)
```
I think this doesn't consider labware overlaps. You can use the utility `getMaxPoolCount` I created

ncdiehl11 commented on protocol-designer/src/components/organisms/LabwareCard/index.tsx:
```diff
@@ -39,22 +42,29 @@ import type { ThunkDispatch } from '/protocol-designer/types'
+        ? deckSetupLabware[id].def.allowedRoles?.includes('adapter')
+        : !deckSetupLabware[id].def.allowedRoles?.includes('adapter'))
   )
+  const isOnHopper = labware.stack.includes(HOPPER_STACKER_LOCATION)
```
isn't there a utility `getIsOnHopper`?

jerader replied:
ya lmfao i think i made it too.

jerader replied:
oh its a utility for the slot being a hopper instead of based off the stack i think

ncdiehl11 commented on protocol-designer/src/components/organisms/SelectLabwareModal/utils.ts:
```diff
@@ -1,7 +1,16 @@
+
+export const getHopperStackLimit = (zHeight: number): number => {
+  return Math.floor(getModuleMaxFillHeight(FLEX_STACKER_MODULE_V1) / zHeight)
+}
```
see my comment above about this not taking into account the overlap

jerader replied:
oof ya i need to refactor all usages of this util

ncdiehl11 commented on protocol-designer/src/components/organisms/SlotDetailsContainer/index.tsx:
```diff
@@ -107,7 +107,11 @@ export function SlotDetailsContainer(
     fullStackFromLabwares.forEach(id => {
       if (deckSetupLabwares[id] != null) {
-        labwares.push(nickNames[id])
+        if (isSlotAHopper) {
+          labwares.push(deckSetupLabwares[id].def.metadata.displayName)
+        } else {
+          labwares.push(nickNames[id])
+        }
```
sorry not sure I'm following this— why do we need to split this logic?

jerader replied:
haha i'll leave a comment. if you add a bunch of duplicate labware, they get nickNames autogenerated and then it creates a bunch of labware items in the slot details causing the glitchiness. so this is a hacky way to have all the labware listed in 1 labware item

ncdiehl11 commented on protocol-designer/src/pages/Designer/DeckSetup/DeckSetupToolbox.tsx:
```diff
@@ -166,8 +165,7 @@ export function DeckSetupToolbox(
     //  NOTE: labware on the Flex Stacker shuttle is not on any module ;)
-    // TODO (nd: 12/11/2025): refactor how we create labware stacks in the hopper
-    if (hasModule) {
+    if (hasModule && !isOnShuttle) {
```
don't we want to create the labware on top of the module, even if it's on the shuttle per our convo? I think we can pass a flag regarding if it's on a hopper or not to make sure we push the `'hopper'` constant when creating the stack. I can address this in a followup if you'd like

jerader replied:
hmmm i'm a bit worried about what happens when you import a protocol, will the labware stack be right on the shuttle? is it ok if we refactor this in a follow up?

ncdiehl11 commented on protocol-designer/src/step-forms/actions/thunks.ts:
I think we can simplify this logic a bunch if we go with allowing hopper groups' labwares' respective stacks to ignore those labware above/below them. If we want to keep in line with the API (as I understand it), the hopper module state should have the responsibility of keeping track of the organization of labware groups inside of it. 

If we go that route, I think we can essentially just wrap the existing logic in a for loop up to the `amount` of 
labware groups

jerader replied:
okay i did a thing, hopefully it works haha

ncdiehl11 reviewed (commented):
Left some comments/questions, but looks good.

jerader committed:
address some feedback

jerader committed:
refactoring

ncdiehl11 committed:
refactor(protocol-designer): re-think handling of stacker in `createContainerAboveModule`

This PR hanldes the logic of creating a container above a stacker, both for loading one labware group on the shuttle, and for loading multiple labware groups in the hopper.

ncdiehl11 reviewed (approved):
I'm comfortable with merging this with the known to-do's being addressed in followups (some architectural discussion still needs to be had with the team). Just a note to resolve the broken test file before merging ✊

jerader committed:
fix test

jerader committed:
fix prettier
